btree/btree.c

```diff
@@ -3038,42 +3293,75 @@ static void generic_move(struct nd *src, struct nd *tgt, enum direction dir,
 	bnode_fix(tgt);
 }
 
-static void ff_capture(struct slot *slot, struct m0_be_tx *tx)
+static void ff_capture(struct slot *slot, int cr, struct m0_be_tx *tx)
 {
-	struct ff_head   *h     = ff_data(slot->s_node);
-	struct m0_be_seg *seg   = slot->s_node->n_tree->t_seg;
-	m0_bcount_t       hsize = sizeof(*h) - sizeof(h->ff_opaque);
+	struct ff_head      *h        = ff_data(slot->s_node);
+	struct ff_dir_entry *de       = (struct ff_dir_entry *)(h + 1);
+	struct m0_be_seg    *seg      = slot->s_node->n_tree->t_seg;
+	m0_bcount_t          hsize    = sizeof(*h) - sizeof(h->ff_opaque);
+	int                  crc_size = 0;
+	int                  kv_cap_size;
 
-	/**
-	 *  Capture starting from the location where new record may have been
-	 *  added or deleted. Capture till the last record. If the deleted
-	 *  record was at the end then no records need to be captured only the
-	 *  header modifications need to be persisted.
-	 */
-	if (h->ff_used > slot->s_idx) {
-		void *start_key        = ff_key(slot->s_node, slot->s_idx);
-		void *last_val         = ff_val(slot->s_node, h->ff_used - 1);
-		int   rec_modify_count = h->ff_used - slot->s_idx;
-		int   vsize            = ff_valsize(slot->s_node);
-		int   crc_size         = 0;
-		int   krsize;
-		int   vrsize;
+	if (h->ff_level == 0 &&
+	    ff_crctype_get(slot->s_node) == M0_BCT_BTREE_ENC_RAW_HASH)
+		crc_size = CRC_VALUE_SIZE;
 
-		if (h->ff_level == 0 &&
-		    ff_crctype_get(slot->s_node) == M0_BCT_BTREE_ENC_RAW_HASH)
-			crc_size = CRC_VALUE_SIZE;
+	if (IS_EMBEDDED_INDIRECT(slot->s_node)) {
+		int de_nr   = h->ff_max_recs;
+		int de_size = sizeof(struct ff_dir_entry) * de_nr;
 
-		krsize     = h->ff_ksize * rec_modify_count;
-		vrsize     = (vsize + crc_size) * rec_modify_count;
+		if (cr == CR_ALL) {
+			kv_cap_size = h->ff_used *
+				      (h->ff_ksize + ff_valsize(slot->s_node) +
+				       crc_size);
 
-		M0_BTREE_TX_CAPTURE(tx, seg, start_key, krsize);
-		M0_BTREE_TX_CAPTURE(tx, seg, last_val, vrsize);
-	} else if (h->ff_opaque == NULL)
+			M0_BTREE_TX_CAPTURE(tx, seg, &de[0], de_size);
+			M0_BTREE_TX_CAPTURE(tx, seg, ff_key(slot->s_node, 0),
+					    kv_cap_size);
+		} else if (cr == CR_NONE) {
+			if (h->ff_max_recs != 0)
+				M0_BTREE_TX_CAPTURE(tx, seg, &de[0], de_size);
+
+			if (h->ff_opaque == NULL)
+				hsize += sizeof(h->ff_opaque);
+		} else {
+			kv_cap_size = h->ff_ksize + ff_valsize(slot->s_node) +
+				      crc_size;
+
+			M0_BTREE_TX_CAPTURE(tx, seg, &de[0], de_size);
```

Comment:
we can avoid capturing the whole directory in this case, instead of that we should capture the directory from i to h->ff_used index.